pkg/check/localpinning/remote_chunks_found.go

```diff
@@ -0,0 +1,140 @@
+package localpinning
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/beekeeper/pkg/bee"
+	"github.com/ethersphere/beekeeper/pkg/beeclient/api"
+	"github.com/ethersphere/beekeeper/pkg/random"
+)
+
+// CheckRemoteChunksFound uploads some chunks to one node, pins them on another
+// node (which does not have them locally), and then check that they are now
+// available locally pinned on that node.
+func CheckRemoteChunksFound(c *bee.Cluster, o Options) error {
+	ctx := context.Background()
+	rnd := random.PseudoGenerator(o.Seed)
+	fmt.Printf("Seed: %d\n", o.Seed)
+
+	ng := c.NodeGroup(o.NodeGroup)
+	overlays, err := ng.Overlays(ctx)
+	if err != nil {
+		return err
+	}
+
+	size := (o.StoreSize / o.StoreSizeDivisor) * swarm.ChunkSize // size in bytes
+	buf := make([]byte, size)
+	_, err = rnd.Read(buf)
+	if err != nil {
+		return fmt.Errorf("rand buffer: %w", err)
+	}
+
+	addrs, err := addresses(buf)
+	if err != nil {
+		return err
+	}
+
+	sortedNodes := ng.NodesSorted()
+	pivot := rnd.Intn(c.Size())
+	pivotNode := sortedNodes[pivot]
+	ref, err := ng.Node(pivotNode).UploadBytes(ctx, buf, api.UploadOptions{Pin: true})
+	if err != nil {
+		return fmt.Errorf("node %s: upload bytes: %w", pivotNode, err)
+	}
+
+	fmt.Printf("uploaded and pinned %d bytes (%d chunks) with hash %s to node %s: %s\n", size, len(addrs), ref.String(), pivotNode, overlays[pivotNode].String())
+
+	// allow the nodes to sync and do some GC
+	time.Sleep(15 * time.Second)
+
+	for name := range overlays {
+		if name == pivotNode {
+			fmt.Printf("Node %s: pivot node, skipping it\n", name)
+			continue
+		}
+
+		nodeClient := ng.Node(name)
+
+		fmt.Printf("Node %s: removing expected chunks\n", name)
+
+		for _, a := range addrs {
+			err := nodeClient.RemoveChunkWithAddress(ctx, a)
```

Comment:
a point to consider, that we might be deleting a chunk which falls under the node's area of responsibility. with a large enough network size this could potentially fail. with a small network it should be OK since the node can still reach the uploader node. just a sidenote, that if we see flakes, this could be the possible reason.

Reply:
The `DELETE` operation should never fail under normal conditions, even when the node does not have the chunk (why I left code that was commented above). The `RemoveChunkWithAddress` only relates to chunks that are locally present on the node.

At worst case scenario (under normal operations) we should get `InternalServerError` that could happen when the first chunk is checked if it is present locally, and then by the time chunk remove operation starts (i.e. `storage.ModeSetRemove`) the chunk was already removed by some other process. We should probably add check in Bee there in the case when the error was `ErrNotFound` to again just return `OK` response.

Reply:
I meant that the flake could come up from a certain edge case where we delete all chunks from a node, but some chunks that should be replicated to other nodes did not (due to asymmetric neighborhoods), hence resulting in some chunks being lost forever, then, once you initiate the pinning process, the node might not be able to reach, download and pin some chunks, resulting in a possible flake.